data-structures/trees/simple-conc/BalanceConc.scala

```diff
@@ -52,24 +78,91 @@ object BalanceConc:
           else r(i - l.size)
     }.ensuring(_ == t.toList(i))
 
+    def contains(v: T): Boolean = {
+      t match
+        case Empty() => false
+        case Leaf(x) => v == x
+        case Node(l, r, _, _) =>  l.contains(v) || r.contains(v)
+    }.ensuring(_ == t.toList.contains(v))
+
     def isBalanced: Boolean = {
       t match {
         case Node(l, r, _, _) =>
           -1 <= l.height - r.height && l.height - r.height <= 1 &&
-          l.isBalanced && r.isBalanced
+          l.isBalanced && r.isBalanced && t.size > 1 // to avoid needlessly deep nodes otherwise the asymptotic complexity can be arbitrary large
         case _ => true
       }
     }
-    
+    def head: T = {
+      require(!t.isEmpty)
+      t match
+        case Leaf(x) => x
+        case Node(l, r, _, _) if l.isEmpty => r.head
+        case Node(l, r, _, _) => l.head
+    }.ensuring(res => res == t.toList.head)
+
+    def last: T = {
+      require(!t.isEmpty)
+      t match
+        case Leaf(x) => x
+        case Node(l, r, _, _) if r.isEmpty => l.last
```

Comment:
Is this case still needed with the invariant?

Reply:
again, we can/should take the invariant as a precondition, or specialize those methods

Reply:
I made the change so that head and tail expect a balance conc to avoid this.